File: HW/HW 8/Solution.java
/**
 * Christopher Oh
 * SJSU CS 146
 * HW 8 - Longest Palindrome
 */

import java.util.HashMap;

public class Solution {
    public Solution() {}

    public int longestPalindrome (String s) {
        int length = 0;
        HashMap<Character, Integer> freq = new HashMap<>(); //letter frequency
        
        for (int i = 0; i < s.length(); i++) {
            if (freq.containsKey(s.charAt(i))) { //map contains character
                freq.put(s.charAt(i), freq.get(s.charAt(i)) + 1);
                if (freq.get(s.charAt(i)) % 2 == 0) { //character has appeared an even amount
                    freq.put(s.charAt(i), 0);
                    length = length + 2;
                }
            } else {
                freq.put(s.charAt(i), 1);
            }
        }

        if (freq.containsValue(1)) length++;

        return length;
    }
}

File: HW/HW 8/Solution.py
# Christopher Oh
# SJSU CS 146
# HW 8 - Longest Palindrome


class Solution:
    def __init__(self):
        print("HW 8")

    def longestPalindrome(self, s):
        length = 0
        freq = dict()

        for letter in s:
            if letter in freq:
                freq[letter] = freq[letter] + 1
                if freq[letter] % 2 == 0:
                    freq[letter] = 0
                    length = length + 2
            else:
                freq[letter] = 1

        for num in freq.values():
            if num == 1:
                length = length + 1
                break

        return length


